internal/config/client.go

```diff
@@ -0,0 +1,99 @@
+package config
+
+import (
+	"context"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/cluster"
+)
+
+// NewCachingClient is an alternative implementation of controller-runtime's
+// default client for manager.Manager.
+// The only difference is that this implementation sets `CacheUnstructured` to `true` to
+// cache unstructured objects.
+func NewCachingClient(cache cache.Cache, config *rest.Config, options client.Options, uncachedObjects ...client.Object) (client.Client, error) {
+	c, err := client.New(config, options)
+	if err != nil {
+		return nil, err
+	}
+
+	return client.NewDelegatingClient(client.NewDelegatingClientInput{
+		CacheReader:       cache,
+		Client:            c,
+		UncachedObjects:   uncachedObjects,
+		CacheUnstructured: true,
+	})
+}
+
+func NewClient(readOnly bool) cluster.NewClientFunc {
+	return func(cache cache.Cache, config *rest.Config, options client.Options, uncachedObjects ...client.Object) (client.Client, error) {
+		c, err := NewCachingClient(cache, config, options, uncachedObjects...)
+		if readOnly {
+			c = &ReadOnlyClient{client: c}
+		}
+		return c, err
+	}
+}
+
+// ReadOnlyClient is a client decorator that ensures no write operations are possible.
+type ReadOnlyClient struct {
+	client client.Client
+}
+
+func (c ReadOnlyClient) GetReadWriteClient() client.Client {
```

Comment:
Maybe add a comment explaining why we have this function. Or can we factor it out instead? Pass a normal client to the validator that needs it, so we don't need this anymore?

Alternatively, `Create()` (line 58) could have special-case code for SubjectAccessReview objects?

Reply:
I prefer to allow all components to use the same client. Was thinking about special handling for SAR in the `Create` func. I definitely prefer that over passing selective client around, and will give that a shot....

Reply:
👍